### DISCUSSION TITLE
Add plugin to sort imports consistently

### DISCUSSION THREAD

**Discussion by kiliman** [epicweb-dev/epic-stack#245] [Ideas]
Since you've added a Prettier plugin to sort Tailwind classes, have you considered adding a plugin to sort imports as well? This will ensure consistency, especially when modifying a file for a PR. 

I used to have the setting: `source.organizeImports` enabled in VS Code. This works great to keep my imports nice and tidy. However, it would cause spurious changes when editing code that didn't have this enabled.

~~With the plugin [`@trivago/prettier-plugin-sort-imports`](https://github.com/trivago/prettier-plugin-sort-imports) installed, everyone will have the same formatting.~~

Use an `eslint` plugin for sorting imports.

**Comment by kentcdodds** [comment]
While changing the order of CSS class names doesn't affect the runtime/specificity of those css classes, the import order *does* have semantic meaning. If your module has side-effects then you need to make sure that module is imported first. We have such a situation here: https://github.com/epicweb-dev/epic-stack/blob/2c13e72ea7f7b04d2be722950488de604ccb62fc/tests/setup/setup-test-env.ts#L1

I suppose we could just add a `// prettier-ignore` comment there, but that's what makes me hesitate. I'm not super excited about formatters changing the semantics of my code 😬

**Reply by kiliman** [reply]
You can configure the sorting rules, so if you want, you can have the local imports first. Perhaps add a prefix to ensure those are always first, like `'./!setup-env-vars.ts'`
```js
// sort ./!* files first, then normal modules, followed by local files
importOrder: ['^[.]\\/!', '<THIRD_PARTY_MODULES>', '^[.]\\/'],
```

Although I prefer to avoid relying on side-effects, at least with this naming convention, it will signal to the developer that this is a special file.

<img width="812" alt="image" src="https://github.com/epicweb-dev/epic-stack/assets/47168/dc0e1f94-7dbc-4223-8b9b-f9a34efdb374">

**Reply by kentcdodds** [reply]
I dunno. Feels like that violates the principle of least surprise to me. I just don't think it's worth it.

**Comment by skjelmo** [comment]
Maybe it could be worthwhile [adding an example](https://github.com/epicweb-dev/epic-stack/blob/main/docs/examples.md) @kiliman?

Thank you both for teaching me something new through this discussion. I didn't even consider imports with side-effects or the semantic meaning of the import order when sorting imports myself. 

I added import sorting as one of the first things, since the imports weren't alphabetically sorted or consistent.

> e.g. `@remix-run/node` in `app/utils/misc.tsx` is second last, while in `app/entry.server.tsx` it is first. 

I still don't get if I'm missing out on some deliberate semantic meaning in the import order, or if there are too many undeliberate inconsistencies for it to be valuable. Maybe I'm missing something? I subjectively feel it's worth the cost of adding a prettier plugin to reduce the mental overhead of an inconsistent import order. It also seems hard to uphold a deliberate semantic import order without prettier. 

I fully understand that you are hesitant to add it to the Epic Stack. I still believe many would benefit from an example. Maybe with a few pointers to tradeoffs to consider.

**Comment by MichaelDeBoey** [comment]
Not sure if we should use a formatter to sort imports tbh as for me a formatter is about formatting your code and sorting imports would be more a task for a linter imo
So [`eslint-plugin-import`](https://github.com/import-js/eslint-plugin-import)'s [`order` rule](https://github.com/import-js/eslint-plugin-import/blob/HEAD/docs/rules/order.md) would be my go-to in this case

**Reply by kentcdodds** [reply]
I may be open to that actually. Because having consistent imports can be nice and it looks like that rule does account for bare imports properly. Anyone want to work on adding it?

**Reply by kiliman** [reply]
`npm i -D eslint-plugin-import`

```diff
// eslintrc.cjs
...
	rules: {
		'@typescript-eslint/consistent-type-imports': [
			'warn',
			{
				prefer: 'type-imports',
				disallowTypeAnnotations: true,
				fixStyle: 'inline-type-imports',
			},
		],
		'@typescript-eslint/no-duplicate-imports': 'warn',
+		'import/order': [
+			'warn',
+			{ alphabetize: { order: 'asc', caseInsensitive: true } },
+		],
	},
```
Optional. eslint --fix on save
```json
// settings.json
  "editor.codeActionsOnSave": {
    "source.fixAll.eslint": true
  },
```

Aside from alphabetizing, I kept the default options. You can be more explicit as needed.

You'll probably want to do one big commit after updating the sort as it will probably affect every file.

**Reply by kentcdodds** [reply]
Thanks! I've implemented this: https://github.com/epicweb-dev/epic-stack/commit/a4a152feb48436f425bcaf06547b5f3a0aa21d08

Two things it doesn't cover that I wish it did:

```ts
import { type DataFunctionArgs } from '@remix-run/node'
import { json, redirect } from '@remix-run/node'

// I want this to auto-fix to

import { json, redirect, type DataFunctionArgs } from '@remix-run/node'
```

And:

```ts
import { type DataFunctionArgs, json, redirect } from '@remix-run/node'

// I want this to auto-fix to:

import { json, redirect, type DataFunctionArgs } from '@remix-run/node'
```

It's possible I'm just doing it wrong. So tips are welcome.

**Reply by kiliman** [reply]
Make these changes:
```diff
	rules: {
		'@typescript-eslint/consistent-type-imports': [
			'warn',
			{
				prefer: 'type-imports',
				disallowTypeAnnotations: true,
				fixStyle: 'inline-type-imports',
			},
		],
		'import/order': [
			'warn',
			{ alphabetize: { order: 'asc', caseInsensitive: true } },
		],
+		'sort-imports': [
+			'warn',
+			{
+				ignoreDeclarationSort: true,
+			},
+		],
		'import/no-duplicates': ['warn', { 'prefer-inline': true }],
	},
```

`sort-imports` will sort the import members alphabetically, with the types coming first. https://stackoverflow.com/a/72769273/293609

I moved `import/no-duplicates` to the bottom and it appears to collapse the duplicate imports.

**Reply by kentcdodds** [reply]
> with the types coming first.

That's unfortunate, because I like them coming last :( That's what VSCode's "organize imports" does.

**Reply by kentcdodds** [reply]
Thanks for the tip on the no-duplicates though, sadly that didn't work for me :(

**Reply by kiliman** [reply]
That's interesting. It's working for me. Here's me doing Cmd+S (save).

As for sorting `type` at the end, I'm still looking at options.

https://github.com/epicweb-dev/epic-stack/assets/47168/5bd73dce-b4d6-466b-b229-e9273c09bdb8

**Reply by kiliman** [reply]
Ok, I found another plugin that does the same as the others combined. I was able to patch it so it will sort the types at the end. It even sorts `exports`.

```
npm i -D eslint-plugin-simple-import-sort

// .eslintrc.cjs
`simple-import-sort/imports': `warn',
'simple-import-sort/exports': 'warn',
```

This patch moves the sort by `kind` to the top and reverses the order (making sure `type` comes *after* `value`).

`eslint-plugin-simple-import-sort+10.0.0.patch`
```diff
diff --git a/node_modules/eslint-plugin-simple-import-sort/shared.js b/node_modules/eslint-plugin-simple-import-sort/shared.js
index 60c92d9..a9c45e8 100644
--- a/node_modules/eslint-plugin-simple-import-sort/shared.js
+++ b/node_modules/eslint-plugin-simple-import-sort/shared.js
@@ -730,6 +730,11 @@ function sortImportExportItems(items) {
 function sortSpecifierItems(items) {
   return items.slice().sort(
     (itemA, itemB) =>
+      // Sort type specifiers after regular ones.
+      compare(
+        getImportExportKind(itemB.node),
+        getImportExportKind(itemA.node)
+      ) ||
       // Compare by imported or exported name (external interface name).
       // import { a as b } from "a"
       //          ^
@@ -745,11 +750,6 @@ function sortSpecifierItems(items) {
       // export { b as a }
       //          ^
       compare(itemA.node.local.name, itemB.node.local.name) ||
-      // Then put type specifiers before regular ones.
-      compare(
-        getImportExportKind(itemA.node),
-        getImportExportKind(itemB.node)
-      ) ||
       // Keep the original order if the names are the same. It’s not worth
       // trying to compare anything else, `import {a, a} from "mod"` is a syntax
       // error anyway (but @babel/eslint-parser kind of supports it).
```


https://github.com/epicweb-dev/epic-stack/assets/47168/039541be-967c-417f-807b-1c8af601843d

**Reply by kentcdodds** [reply]
Nice! I think I'd rather avoid patching, but if anyone wants to make a PR and get that in the official lib, I'd love to use it.

**Reply by kiliman** [reply]
BTW: This issue discusses import sorting https://github.com/lydell/eslint-plugin-simple-import-sort/issues/124. The behavior in v10 is to sort import specifiers alphabetically, even types. So imports would sort like:
```ts
import {
	type ActionArgs,
	json,
	type LoaderArgs,
	redirect,
} from '@remix-run/node'
```
The main rationale is consistency with other tooling as well as minimizing diff changes (simply add or remove type specifier without resorting).

Using the patch will sort types at the end. Considering the issue above, not sure if they would accept a PR to add a setting for this.

**Reply by jacobparis** [reply]
One possible alternative is this lovely plugin that lets you register local Eslint rules, so you can add any Epic stack opinionated rules right into the codebase without needing to publish/patch 

I've used this in the past to enforce capitalization styleguides in locale files via a local eslint plugin

https://www.npmjs.com/package/eslint-plugin-local-rules

**Reply by kiliman** [reply]
So if I'm not misunderstanding, that plugin allows you to create your own *local* plugin, essentially. Unfortunately, in this case, it would require essentially copying the original plugin with our custom change vs `patch-package`, since we need to update the internal sort algorithm.

With that being said, it looks like even the TypeScript team is adopting `eslint-plugin-simple-sort` with VS Code Organize Imports and auto-imports eventually doing the same thing (default to sort ordinal).

So perhaps it may be best to just go with the flow and let the plugin do its thing.

https://github.com/microsoft/TypeScript/pull/52090
https://github.com/microsoft/TypeScript/pull/52115
https://github.com/microsoft/TypeScript/issues/52820
